Public/Src/Tools/SymbolDaemon/SymbolDaemon.cs

```diff
@@ -0,0 +1,424 @@
+            m_logger.Info(I($"Using {nameof(SymbolConfig)}: {JsonConvert.SerializeObject(symbolConfig)}"));
+
+            // if no ISymbolServiceClient has been provided, create VsoSymbolClient using the provided SymbolConfig
+            m_symbolServiceClientTask = symbolServiceClientTask ?? Task.Run(() => (ISymbolClient)new VsoSymbolClient(m_logger, symbolConfig));
+
+            m_symbolIndexer = new SymbolIndexer(SymbolAppTraceSource.SingleInstance);
+        }
+
+        internal static void EnsureCommandsInitialized()
+        {
+            Contract.Assert(Commands != null);
+
+            // these operations are quite expensive, however, we expect to call this method only once per symbol request, so it should cause any perf downgrade
+            var numCommandsBase = typeof(ServicePipDaemon.ServicePipDaemon).GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static).Where(f => f.FieldType == typeof(Command)).Count();
+            var numCommandsSymbolD = typeof(SymbolDaemon).GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static).Where(f => f.FieldType == typeof(Command)).Count();
+
+            if (Commands.Count != numCommandsBase + numCommandsSymbolD)
+            {
+                Contract.Assert(false, $"The list of commands was not properly initialized (# of initialized commands = {Commands.Count}; # of ServicePipDaemon commands = {numCommandsBase}; # of SymbolDaemon commands = {numCommandsSymbolD})");
+            }
+        }
+
+        /// <summary>
+        /// Synchronous version of <see cref="CreateAsync"/>
+        /// </summary>
+        public IIpcResult Create()
+        {
+            return CreateAsync().GetAwaiter().GetResult();
+        }
+
+        /// <summary>
+        /// Creates a symbol request.
+        /// </summary>
+        public async Task<IIpcResult> CreateAsync()
+        {
+            // TODO(olkonone): add logging
+            Request createRequestResult;
+
+            try
+            {
+                createRequestResult = await InternalCreateAsync();
+            }
+            catch (Exception e)
+            {
+                return new IpcResult(IpcResultStatus.GenericError, e.DemystifyToString());
+            }
+
+            // get and return RequestId ?
+            return IpcResult.Success(I($"Symbol request '{RequestName}' created (assigned request ID: '{createRequestResult.Id}')."));
+        }
+
+        /// <summary>
+        /// Indexes the files and adds symbol data to the request.
+        /// </summary>       
+        public async Task<IIpcResult> AddSymbolFilesAsync(List<SymbolFile> files)
+        {
+            var result = await EnsureFilesAreIndexedAsync(files);
+
+            if (!result.Success)
+            {
+                return new IpcResult(IpcResultStatus.ExecutionError, result.Exception.DemystifyToString());
+            }
+
+            var addFileTasks = files.Select(f => AddSymbolFileAsync(f));

```

Comment:
>AddSymbolFileAsync [](start = 49, length = 18)

Instead of doing it per file, we can batch it, i guess. UploadFile method call will be still per-file; but CreateRequestDebugEntries support batching

Reply:
Yes, we can first collect all the `DebugEntries` and call `CreateRequestDebugEntries`. Then get a set of files that need to be uploaded and upload them (one by one - it's the only api we have). Finally, we'll need to call `CreateRequestDebugEntries` again for all the entries that failed in the first call (note: there might be more entries than files)

---
In reply to: [314220424](https://github.com/microsoft/BuildXL/pull/727#discussion_r314220424) [](ancestors = 314220424)

Reply:
I'll add this change in 'caching' PR.

---
In reply to: [320083410](https://github.com/microsoft/BuildXL/pull/727#discussion_r320083410) [](ancestors = 320083410,314220424)